### PR TITLE
Enable faulthandler for distributed tests.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54818 Use faulthandler to dump traceback of timed out processes in unit tests.
* **#54531 Enable faulthandler for distributed tests.**

Enabling faulthandler will intercept signals like SIGSEGV, SIGFPE,
SIGABRT, SIGBUS and SIGKILL and dump the entire python traceback before the
process goes down.

This can help us in debugging flaky tests where a process crashes and we need
to debug what happened.

faulthandler trace looks like this:
```
Fatal Python error: Aborted

Current thread 0x00007f04d36bb740 (most recent call first):
  File "test/distributed/test_c10d.py", line 4634 in <lambda>
  File "test/distributed/test_c10d.py", line 4610 in _test_nccl_errors_blocking
  File "test/distributed/test_c10d.py", line 4634 in test_nccl_errors_blocking_abort
  .....
```

Differential Revision: [D27271048](https://our.internmc.facebook.com/intern/diff/D27271048/)

### CODE REVIEW

**Review by rohan-varma — APPROVED** [review]
Looks good! I'm assuming this won't produce C++ traces but python trace should be better than what we currently have. Could you paste an example of what we can expect with this change? Also, does it matter whether crash happened in C++ or python?

**Comment by pritamdamania87** [comment]
> Looks good! I'm assuming this won't produce C++ traces but python trace should be better than what we currently have. Could you paste an example of what we can expect with this change? Also, does it matter whether crash happened in C++ or python?

@rohan-varma Had to change the faulthandler to dump to a file instead since `file=sys.stderr` doesn't work in certain environments. Yes it does only dump the python traces and it doesn't matter if the crash happened in C++/python.

File: torch/testing/_internal/common_distributed.py [hunk 1]
```diff
@@ -330,16 +337,19 @@ def _start_processes(self, proc):
...
             parent_conn, child_conn = torch.multiprocessing.Pipe()
+            faulthandler_file = tempfile.NamedTemporaryFile()
             process = proc(
                 target=self.__class__._run,
                 name='process ' + str(rank),
-                args=(rank, self._current_test_name(), self.file_name, child_conn))
+                args=(rank, self._current_test_name(), self.file_name, child_conn, faulthandler_file.name))
             process.start()
-            logging.info('Started process {} with pid {}'.format(rank, process.pid))
+            logger.info('Started process {} with pid {}'.format(rank, process.pid))
```
Comment (H-Huang) [inline torch/testing/_internal/common_distributed.py:349]: nit: can replace str.format with [f-strings](https://docs.python.org/3.6/whatsnew/3.6.html#whatsnew36-pep498). E.g. `f'Started process {rank} with pid {process.pid}'`. Is a bit easier to read and is supposedly faster

Same comment for other strings added in PR.

File: torch/testing/_internal/common_distributed.py [hunk 2]
```diff
@@ -318,6 +321,10 @@ def tearDown(self):
         # processes to prevent an effective file descriptor leak.
         self.processes = []
 
+        # Close all the files
```
Comment (H-Huang) [inline torch/testing/_internal/common_distributed.py:324]: never used faulthandler before, is `faulthandler.disable()` also needed?

**Review by H-Huang — APPROVED** [review]
Great addition!

File: torch/testing/_internal/common_distributed.py (hunk 2 above)
Comment (pritamdamania87) [reply torch/testing/_internal/common_distributed.py:324]: I don't think it is necessary, you need to do it only if you want to disable the signal handlers at some point.

**Pull request closed without merging**